File: Final/backend/schemas/smtp_providers.py
"""
SMTP Provider Configuration Schemas
"""

from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, Field, validator

from models.smtp_providers import SMTPProvider


class ProviderConfigBase(BaseModel):
    """Base schema for provider configuration"""
    provider: SMTPProvider
    name: str = Field(..., min_length=1, max_length=255)
    is_active: bool = True


class ProviderConfigCreate(ProviderConfigBase):
    """Schema for creating a provider configuration"""
    credentials: dict[str, str]
    custom_limits: Optional[dict[str, Any]] = None
    
    @validator("credentials")
    def validate_credentials(cls, v, values):
        provider = values.get("provider")
        
        if provider == SMTPProvider.AWS_SES:
            required = ["access_key_id", "secret_access_key", "region"]
            if not all(field in v for field in required):
                raise ValueError(f"AWS SES requires: {', '.join(required)}")
        
        elif provider == SMTPProvider.SENDGRID:
            if "api_key" not in v:
                raise ValueError("SendGrid requires api_key")
        
        elif provider == SMTPProvider.MAILGUN:
            required = ["api_key", "domain"]
            if not all(field in v for field in required):
                raise ValueError(f"Mailgun requires: {', '.join(required)}")
        
        elif provider == SMTPProvider.CUSTOM:
            required = ["smtp_host", "smtp_port", "smtp_username", "smtp_password"]
            if not all(field in v for field in required):
                raise ValueError(f"Custom SMTP requires: {', '.join(required)}")
        
        return v


class ProviderConfigUpdate(BaseModel):
    """Schema for updating a provider configuration"""
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    is_active: Optional[bool] = None
    
    # Limit updates
    sending_quota_daily: Optional[int] = Field(None, ge=0)
    sending_quota_hourly: Optional[int] = Field(None, ge=0)
    sending_rate_per_second: Optional[float] = Field(None, ge=0)
    max_concurrent_connections: Optional[int] = Field(None, ge=1)
    max_message_size_mb: Optional[int] = Field(None, ge=1)
    max_recipients_per_message: Optional[int] = Field(None, ge=1)
    
    # Warmup configuration
    is_warming_up: Optional[bool] = None
    warmup_daily_increment: Optional[int] = Field(None, ge=10, le=1000)


class ProviderLimits(BaseModel):
    """Provider limit information"""
    daily: dict[str, int]
    hourly: dict[str, int]
    per_second: float
    concurrent_connections: int


class ProviderWarmupInfo(BaseModel):
    """Warmup information"""
    is_warming_up: bool
    current_limit: Optional[int]
    start_date: Optional[datetime]


class ProviderConfigResponse(ProviderConfigBase):
    """Response schema for provider configuration"""
    id: str
    smtp_host: str
    smtp_port: int
    is_verified: bool
    health_status: str
    reputation_score: float
    limits: ProviderLimits
    warmup: ProviderWarmupInfo
    created_at: datetime
    updated_at: datetime
    
    class Config:
        orm_mode = True
        
    @classmethod
    def from_orm(cls, obj):
        """Create from ORM model"""
        return cls(
            id=obj.id,
            provider=obj.provider,
            name=obj.name,
            is_active=obj.is_active,
            smtp_host=obj.smtp_host,
            smtp_port=obj.smtp_port,
            is_verified=obj.is_verified,
            health_status=obj.health_status,
            reputation_score=obj.reputation_score,
            limits=ProviderLimits(
                daily={
                    "limit": obj.get_current_limit("daily"),
                    "used": obj.daily_sent_count,
                    "remaining": obj.get_remaining_quota("daily"),
                },
                hourly={
                    "limit": obj.get_current_limit("hourly"),
                    "used": obj.hourly_sent_count,
                    "remaining": obj.get_remaining_quota("hourly"),
                },
                per_second=obj.sending_rate_per_second,
                concurrent_connections=obj.max_concurrent_connections,
            ),
            warmup=ProviderWarmupInfo(
                is_warming_up=obj.is_warming_up,
                current_limit=obj.warmup_current_limit,
                start_date=obj.warmup_start_date,
            ),
            created_at=obj.created_at,
            updated_at=obj.updated_at,
        )


class ProviderQuotaResponse(BaseModel):
    """Response for quota check"""
    allowed: bool
    checks: dict[str, bool]
    quotas: dict[str, Any]
    reason: Optional[str] = None


class ProviderUsageStats(BaseModel):
    """Provider usage statistics"""
    config_id: str
    provider: SMTPProvider
    name: str
    period_hours: int
    total_sent: int
    total_bounced: int
    total_complained: int
    total_errors: int
    bounce_rate: float
    complaint_rate: float
    average_send_time_ms: Optional[float]
    reputation_score: float
    health_status: str

File: Final/frontend/src/api/smtp-providers.ts
import { apiClient } from '@/http/apiClient';

export interface SMTPProviderOption {
    value: string;
    label: string;
    description: string;
}

export interface ProviderConfigCreate {
    provider: string;
    name: string;
    credentials: Record<string, string>;
    custom_limits?: Record<string, any>;
}

export interface ProviderConfigUpdate {
    name?: string;
    is_active?: boolean;
    sending_quota_daily?: number;
    sending_quota_hourly?: number;
    sending_rate_per_second?: number;
    max_concurrent_connections?: number;
    max_message_size_mb?: number;
    max_recipients_per_message?: number;
    is_warming_up?: boolean;
    warmup_daily_increment?: number;
}

export interface ProviderConfig {
    id: string;
    provider: string;
    name: string;
    is_active: boolean;
    is_verified: boolean;
    health_status: string;
    reputation_score: number;
    smtp_host: string;
    smtp_port: number;
    limits: {
        daily: { limit: number; used: number; remaining: number };
        hourly: { limit: number; used: number; remaining: number };
        per_second: number;
        concurrent_connections: number;
    };
    warmup: {
        is_warming_up: boolean;
        current_limit: number | null;
        start_date: string | null;
    };
    created_at: string;
    updated_at: string;
}

export interface ProviderQuotaCheck {
    allowed: boolean;
    checks: Record<string, boolean>;
    quotas: Record<string, any>;
    reason?: string;
}

export interface ProviderUsageStats {
    config_id: string;
    provider: string;
    name: string;
    period_hours: number;
    total_sent: number;
    total_bounced: number;
    total_complained: number;
    total_errors: number;
    bounce_rate: number;
    complaint_rate: number;
    average_send_time_ms?: number;
    reputation_score: number;
    health_status: string;
}

export const smtpProviderApi = {
    // Get available providers
    async getAvailableProviders(): Promise<SMTPProviderOption[]> {
        const response = await apiClient.get('/smtp-providers/available');
        return response.data;
    },

    // Create a new provider configuration
    async createConfig(data: ProviderConfigCreate): Promise<ProviderConfig> {
        const response = await apiClient.post('/smtp-providers/configs', data);
        return response.data;
    },

    // Get all provider configurations
    async getConfigs(provider?: string, activeOnly = true): Promise<ProviderConfig[]> {
        const params = new URLSearchParams();
        if (provider) params.append('provider', provider);
        params.append('active_only', activeOnly.toString());

        const response = await apiClient.get(`/smtp-providers/configs?${params}`);
        return response.data;
    },

    // Get a specific provider configuration
    async getConfig(configId: string): Promise<ProviderConfig> {
        const response = await apiClient.get(`/smtp-providers/configs/${configId}`);
        return response.data;
    },

    // Update a provider configuration
    async updateConfig(configId: string, data: ProviderConfigUpdate): Promise<ProviderConfig> {
        const response = await apiClient.put(`/smtp-providers/configs/${configId}`, data);
        return response.data;
    },

    // Delete a provider configuration
    async deleteConfig(configId: string): Promise<void> {
        await apiClient.delete(`/smtp-providers/configs/${configId}`);
    },

    // Check quota for a provider
    async checkQuota(configId: string, emailsToSend = 1): Promise<ProviderQuotaCheck> {
        const response = await apiClient.get(
            `/smtp-providers/configs/${configId}/quota?emails_to_send=${emailsToSend}`
        );
        return response.data;
    },

    // Get usage statistics
    async getUsageStats(configId: string): Promise<ProviderUsageStats> {
        const response = await apiClient.get(`/smtp-providers/configs/${configId}/usage`);
        return response.data;
    },

    // Test a provider configuration
    async testConfig(configId: string, testEmail: string): Promise<any> {
        const response = await apiClient.post(`/smtp-providers/configs/${configId}/test`, {
            test_email: testEmail,
        });
        return response.data;
    },

    // Start warmup process
    async startWarmup(configId: string, dailyIncrement = 50): Promise<any> {
        const response = await apiClient.post(
            `/smtp-providers/configs/${configId}/warmup/start?daily_increment=${dailyIncrement}`
        );
        return response.data;
    },

    // Stop warmup process
    async stopWarmup(configId: string): Promise<any> {
        const response = await apiClient.post(`/smtp-providers/configs/${configId}/warmup/stop`);
        return response.data;
    },
};
